packages/slate-react/src/components/editor.js

```diff
@@ -71,140 +72,112 @@ class Editor extends React.Component {
-    this.flushChange()
+  componentDidUpdate(prevProps) {
+    // Increment the updates counter as a baseline.
+    this.tmp.updates++
+
+    // If we've resolved a few times already, and it's exactly in line with
+    // the updates, then warn the user that they may be doing something wrong.
+    if (this.tmp.resolves > 5 && this.tmp.resolves == this.tmp.updates) {
+      logger.warn(
+        'A Slate <Editor> is re-resolving `props.plugins` or `props.schema` on each update, which leads to poor performance. This is often due to passing in a new `schema` or `plugins` prop with each render by declaring them inline in your render function. Do not do this!'
+      )
+    }
+
+    this.change(change => change)
   }
 
   /**
-   * Queue a `change` object, to be able to flush it later. This is required for
-   * when a change needs to be applied to the value, but because of the React
-   * lifecycle we can't apply that change immediately. So we cache it here and
-   * later can call `this.flushChange()` to flush it.
-   *
-   * @param {Change} change
+   * Get event handlers as passed by onEvent
    */
 
-  queueChange = change => {
-    if (change.operations.size) {
-      debug('queueChange', { change })
-      this.tmp.change = change
+  handlers = EVENT_HANDLERS.reduce((obj, handler) => {
+    obj[handler] = event => {
+      this.onEvent(handler, event)
     }
-  }
+    return obj
+  }, {})
 
   /**
-   * Flush a temporarily stored `change` object, for when a change needed to be
-   * made but couldn't because of React's lifecycle.
+   * Perform a change on the editor, passing `...args` to `change.call`.
+   *
+   * @param {Mixed} ...args
    */
 
-  flushChange = () => {
+  change = (...args) => {
     const { change } = this.tmp
+    const lastOperationSize = change.operations.size
+    change.call(...args)
+    debug('onChange', { change })
+
+    // Do not rerun the change if onChange is run already in associateStackAndValue without more following operations
+    if (change.operations.size > lastOperationSize) {
+      this.stack.run('onChange', change, this)
+    }
 
-    if (change) {
-      debug('flushChange', { change })
-      delete this.tmp.change
-      this.props.onChange(change)
+    // After we update the change with props.onChange; do not re-update in second
+    // cycle if no other operations are added
+    if (change.operations.size === 0 && change.value === this.props.value) {
+      return
     }
+
+    this.updateChange(change)
   }
 
   /**
-   * Perform a change on the editor, passing `...args` to `change.call`.
+   * On change.
    *
-   * @param {Mixed} ...args
+   * @param {Change} change
    */
 
-  change = (...args) => {
-    const change = this.value.change().call(...args)
-    this.onChange(change)
+  onChange = change => {
+    debug('onChange', { change })
+    const { value } = change
+    if (value == this.props.value && change.operations.size === 0) return
+    this.stack.run('onChange', change, this)
```

Comment:
Thanks for re-adding this `onChange` method. 

We've now got another issue with this logic though, which is that there are three different places that `stack.run('onChange'` is called from, which I don't think is good because it makes things hard to follow.

Right now, it seems like one of the reasons for the duplication is that in the `componentDidMount` we're doing `this.change(change => change)` to actually flush any existing changes. This feels like too much magic, it isn't clear. And if that line was refactored to be more clear and not depend on the same code path, I think it would allow us to remove the duplcation.

For example (pseudocode):

```js
if (this.tmp.change) {
  this.flushChange()
}
```

Ideally `editor.change` should be reduced to:

```js
change = (...args) => {
  const { value } = this
  const change = value.change()
  change.call(...args)
  this.onChange(change)
}
```

Otherwise it probably means we're doing something too confusing.

Reply:
Hi, @ianstormtaylor .  How about 3257e3e?  The problem is that `onChange(change)` can take an arbitrary change that not restricted to the inner state of the editor.  So we have to decide whether to run `stack.run(`onChange`)` in difference places.  One for inner change state `this.tmp.change` in `this.change`, the other for arbitrary change in `this.onChange`